nanodbc/nanodbc.cpp

```diff
@@ -3161,6 +3165,37 @@ inline void result::result_impl::get_ref_impl<std::vector<std::uint8_t>>(
     throw type_incompatible_error();
 }
 
+template <>
+void result::result_impl::get_ref_from_string_column(short column, string::value_type& result) const
+{
+    bound_column& col = bound_columns_[column];
+    const char* s = col.pdata_ + rowset_position_ * col.clen_;
+    switch (col.ctype_)
+    {
+    case SQL_C_CHAR:
+        result = (string::value_type) * (const char*)(s);
+        return;
+    case SQL_C_WCHAR:
+        result = (string::value_type) * (const SQLWCHAR*)(s);
+        return;
+    }
+    throw type_incompatible_error();
+}
+
+template <class T>
+void result::result_impl::get_ref_from_string_column(short column, T& result) const
+{
+    bound_column& col = bound_columns_[column];
+    if (col.ctype_ != SQL_C_CHAR && col.ctype_ != SQL_C_WCHAR)
+        throw type_incompatible_error();
+    string str;
+    get_ref_impl(col.column_, str);
+    std::string temp;
+    convert(str, temp);
+    std::istringstream strstr(temp);
```

Comment:
I'd suggest to avoid the streams and use the C++11 `stox` functions with a little bit of indirection, e.g.:

```
namespace detail {
template <typename T>
auto from_string(T const& s, double) { return std::stod(s); }
template <typename T>
auto from_string(T const& s, int) { return std::stoi(s); }
template <typename T>
auto from_string(T const& s, long) { return std::stol(s); }
}

template <typename R, typename T>
auto from_string(T const& s) -> R
{
    return detail::from_string(s, R{});
}
```

then the line in `get_ref_from_string_column` above would read:

```
result = from_string<T>(str);
```

Quick test w/ VS2017 suggests the `stoX` are 10-20 faster

Reply:
Yes this is a good idea, it is not only faster, it would also throw an out_of_range exception if the value is to big (sstream simply returns the biggest possible value). The only problem is that there is no function for short and with a simple cast from stoi you would sometimes throw an out_of_range error and sometimes not but return a wrong value.
In c++17 there is `std::from_chars` which also works for short (though the error handling is strange).
Another solution would be boost::lexical_cast.

Reply:
> there is no function for short and with a simple cast from stoi you would sometimes
> throw an out_of_range error and sometimes not but return a wrong value.

Good point. `from_string(T const& s, short)` would have to get a bit of logic checking `std::stoi` return value fits the range of `short`, I think.

> In c++17 there is std::from_chars 

Yes, but AFAIK, nanodbc 2.x line (also current master) sticks to C++14

> Another solution would be boost::lexical_cast.

Boost should be considered optional, and `short` could be handled with that but only if `NANODBC_ENABLE_BOOST=ON`

Reply:
`boost::lexical_cast` is kinda neat, I was following the boost mailing lists as that was going through it's development and review. But nanodbc aims to be as compact as possible without reliance on other libraries, no matter how generally accepted and supported they are. If we can use what's standard then we should try to at least.

Reply:
Ok, i've changed the stringstream conversion to the from_string style (it uses stoll/stoull for integer types, and checks if it is inside the limits of the requested type). Can somebody run the tests, travis/appveyor seems not to be working on this repository?

Reply:
The project is in middle of transition to GitHub organization.

I think we will have to wait with the merge, or perhaps we don't? What we do @lexicalunit ?